File: src/stores/dailyChallenge.ts
import { computed, ref } from 'vue'
import { defineStore } from 'pinia'

import request from '@/utils/request'
import type {
  AnswerDetail,
  QuestionContent0,
  QuestionContent1,
  QuestionContent2,
  QuestionType,
} from '@/stores/combat'

export type DailyChallengeStatus = 'in_progress' | 'dead' | 'completed_exhausted'
export type DailyChallengeUserStatus = 'not_started' | 'in_progress' | 'ended'

export interface DailyChallengeQuestion {
  question_id: number
  floor: number
  question_index: number
  type: QuestionType
  content: QuestionContent0 | QuestionContent1 | QuestionContent2
}

export interface DailyChallengeAnswerResult {
  is_correct: boolean
  detail: Record<string, unknown>
}

export interface DailyChallengeState {
  run_id: number
  day_key: string
  status: DailyChallengeStatus
  current_floor: number
  current_question_index: number
  current_hp: number
  current_enemy_hp: number
  player_max_hp: number
  player_attack: number
  enemy_max_hp: number
  enemy_attack: number
  heal_every_floors: number
  heal_amount: number
  best_floor: number
  best_floor_reached_at: string
  started_at: string
  ended_at: string | null
  question: DailyChallengeQuestion | null
  answer_result: DailyChallengeAnswerResult | null
}

export interface DailyChallengeLeaderboardEntry {
  rank: number
  user_id: number
  username: string
  nickname: string | null
  avatar_url: string | null
  best_floor: number
  best_floor_reached_at: string | null
  started_at: string
  status: DailyChallengeStatus
  is_current_user: boolean
}

export interface DailyChallengeOverview {
  day_key: string
  opens_at: string
  closes_at: string
  seconds_until_reset: number
  user_status: DailyChallengeUserStatus
  today_best_floor: number | null
  active_run: DailyChallengeState | null
  leaderboard: DailyChallengeLeaderboardEntry[]
  current_user_rank: number | null
  current_user_entry: DailyChallengeLeaderboardEntry | null
}

type CountdownTimer = ReturnType<typeof setInterval> | null

function cloneQuestion(question: DailyChallengeQuestion | null | undefined): DailyChallengeQuestion | null {
  if (!question) return null
  return JSON.parse(JSON.stringify(question)) as DailyChallengeQuestion
}

export const useDailyChallengeStore = defineStore('dailyChallenge', () => {
  const overview = ref<DailyChallengeOverview | null>(null)
  const leaderboard = ref<DailyChallengeLeaderboardEntry[]>([])
  const leaderboardDayKey = ref<string | null>(null)
  const currentUserRank = ref<number | null>(null)
  const currentUserEntry = ref<DailyChallengeLeaderboardEntry | null>(null)

  const activeState = ref<DailyChallengeState | null>(null)
  const displayedQuestion = ref<DailyChallengeQuestion | null>(null)
  const answerResult = ref<DailyChallengeAnswerResult | null>(null)
  const pendingState = ref<DailyChallengeState | null>(null)
  const secondsUntilReset = ref(0)

  const isLoadingOverview = ref(false)
  const isLoadingLeaderboard = ref(false)
  const isStartingChallenge = ref(false)
  const isSubmittingAnswer = ref(false)

  const countdownTimer = ref<CountdownTimer>(null)

  const hasActiveRun = computed(() => activeState.value?.status === 'in_progress')
  const canContinueCurrentQuestion = computed(() => answerResult.value !== null)
  const isEnded = computed(
    () =>
      activeState.value?.status === 'dead' || activeState.value?.status === 'completed_exhausted',
  )

  function stopCountdown() {
    if (countdownTimer.value) {
      clearInterval(countdownTimer.value)
      countdownTimer.value = null
    }
  }

  function startCountdown(seconds: number) {
    stopCountdown()
    secondsUntilReset.value = Math.max(0, seconds)
    countdownTimer.value = setInterval(() => {
      secondsUntilReset.value = Math.max(0, secondsUntilReset.value - 1)
    }, 1000)
  }

  function hydrateOverview(data: DailyChallengeOverview) {
    overview.value = data
    leaderboard.value = data.leaderboard ?? []
    leaderboardDayKey.value = data.day_key
    currentUserRank.value = data.current_user_rank
    currentUserEntry.value = data.current_user_entry
    startCountdown(data.seconds_until_reset ?? 0)
  }

  function applyFreshState(state: DailyChallengeState | null) {
    activeState.value = state
    displayedQuestion.value = cloneQuestion(state?.question)
    answerResult.value = null
    pendingState.value = null
  }

  async function refreshLeaderboardInBackground() {
    try {
      await fetchLeaderboard()
    } catch (error) {
      console.error('刷新每日挑战排行榜失败:', error)
    }
  }

  async function fetchOverview() {
    isLoadingOverview.value = true
    try {
      const res = await request.get('/api/daily-challenge/overview', { timeout: 10000 })
      if (res.data.success) {
        const data = res.data.data as DailyChallengeOverview
        hydrateOverview(data)
        if (data.active_run) {
          applyFreshState(data.active_run)
        } else if (!pendingState.value) {
          activeState.value = null
          displayedQuestion.value = null
          answerResult.value = null
          pendingState.value = null
        }
      }
      return res.data
    } finally {
      isLoadingOverview.value = false
    }
  }

  async function fetchLeaderboard() {
    isLoadingLeaderboard.value = true
    try {
      const res = await request.get('/api/daily-challenge/leaderboard', { timeout: 10000 })
      if (res.data.success) {
        const data = res.data.data as {
          day_key: string
          items: DailyChallengeLeaderboardEntry[]
          current_user_rank: number | null
          current_user_entry: DailyChallengeLeaderboardEntry | null
        }
        leaderboard.value = data.items ?? []
        leaderboardDayKey.value = data.day_key
        currentUserRank.value = data.current_user_rank
        currentUserEntry.value = data.current_user_entry
      }
      return res.data
    } finally {
      isLoadingLeaderboard.value = false
    }
  }

  async function startOrResume() {
    isStartingChallenge.value = true
    try {
      const res = await request.post('/api/daily-challenge/start', null, { timeout: 30000 })
      if (res.data.success) {
        const state = res.data.data as DailyChallengeState
        applyFreshState(state)
        if (overview.value) {
          overview.value.active_run = state
          overview.value.user_status = state.status === 'in_progress' ? 'in_progress' : 'ended'
          overview.value.today_best_floor = state.best_floor
        }
      }
      return {
        success: Boolean(res.data.success),
        message: String(res.data.message ?? ''),
        state: res.data.success ? (res.data.data as DailyChallengeState) : null,
      }
    } finally {
      isStartingChallenge.value = false
    }
  }

  async function submitAnswer(payload: {
    selected_option?: string
    selected_sequence?: string[]
    user_input?: string
    is_correct?: boolean
    answer_detail?: AnswerDetail
  }) {
    if (!displayedQuestion.value || isSubmittingAnswer.value) {
      return { success: false, message: '当前没有可提交的题目' }
    }

    isSubmittingAnswer.value = true
    try {
      const res = await request.post('/api/daily-challenge/answer', {
        question_id: displayedQuestion.value.question_id,
        ...payload,
      }, { timeout: 30000 })

      if (res.data.success) {
        const next = res.data.data as DailyChallengeState
        activeState.value = next
        answerResult.value = next.answer_result
        pendingState.value = next
        if (overview.value) {
          overview.value.active_run = next.status === 'in_progress' ? next : null
          overview.value.user_status = next.status === 'in_progress' ? 'in_progress' : 'ended'
          overview.value.today_best_floor = next.best_floor
        }
        void refreshLeaderboardInBackground()
      }

      return {
        success: Boolean(res.data.success),
        message: String(res.data.message ?? ''),
      }
    } finally {
      isSubmittingAnswer.value = false
    }
  }

  function continueAfterAnswer() {
    const next = pendingState.value
    if (!next) return
    displayedQuestion.value = cloneQuestion(next.question)
    answerResult.value = null
    pendingState.value = null
  }

  function clearActiveRun() {
    activeState.value = null
    displayedQuestion.value = null
    answerResult.value = null
    pendingState.value = null
  }

  return {
    overview,
    leaderboard,
    leaderboardDayKey,
    currentUserRank,
    currentUserEntry,
    activeState,
    displayedQuestion,
    answerResult,
    pendingState,
    secondsUntilReset,
    isLoadingOverview,
    isLoadingLeaderboard,
    isStartingChallenge,
    isSubmittingAnswer,
    hasActiveRun,
    canContinueCurrentQuestion,
    isEnded,
    stopCountdown,
    fetchOverview,
    fetchLeaderboard,
    startOrResume,
    submitAnswer,
    continueAfterAnswer,
    clearActiveRun,
  }
})

File: app/models/daily_challenge.py
from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Any, Optional

from sqlalchemy import JSON
from sqlalchemy import Column, String, UniqueConstraint
from sqlmodel import Field, SQLModel

from sqlalchemy.dialects.postgresql import JSONB


def _json_type():
    return JSONB().with_variant(JSON(), "sqlite")


class DailyChallengeRunStatus(str, Enum):
    IN_PROGRESS = "in_progress"
    DEAD = "dead"
    COMPLETED_EXHAUSTED = "completed_exhausted"


class DailyChallengeQuestionType(str, Enum):
    CONTEXT_GUESS = "context_guess"
    CLOZE_TEST = "cloze_test"
    KEYWORD_TRANSLATION = "keyword_translation"


class DailyChallengeDay(SQLModel, table=True):
    __tablename__ = "daily_challenge_day"

    id: Optional[int] = Field(default=None, primary_key=True)
    day_key: str = Field(index=True, unique=True, max_length=32)
    opens_at: datetime
    closes_at: datetime
    source_tags: list[str] = Field(default_factory=list, sa_column=Column(_json_type()))
    config_snapshot: dict[str, Any] = Field(default_factory=dict, sa_column=Column(_json_type()))
    created_at: datetime


class DailyChallengeFloorQuestion(SQLModel, table=True):
    __tablename__ = "daily_challenge_floor_question"
    __table_args__ = (
        UniqueConstraint("day_id", "floor", "question_index", name="uq_daily_floor_question_position"),
    )

    id: Optional[int] = Field(default=None, primary_key=True)
    day_id: int = Field(foreign_key="daily_challenge_day.id", index=True)
    floor: int = Field(index=True)
    question_index: int = Field(default=1)
    question_id: int = Field(foreign_key="question.id")
    question_type: str = Field(sa_column=Column(String(100)))
    word_ids: list[int] = Field(default_factory=list, sa_column=Column(_json_type()))
    created_at: datetime


class DailyChallengeRun(SQLModel, table=True):
    __tablename__ = "daily_challenge_run"
    __table_args__ = (
        UniqueConstraint("day_id", "user_id", name="uq_daily_run_day_user"),
    )

    id: Optional[int] = Field(default=None, primary_key=True)
    day_id: int = Field(foreign_key="daily_challenge_day.id", index=True)
    user_id: int = Field(foreign_key="user.id", index=True)
    status: DailyChallengeRunStatus = Field(
        default=DailyChallengeRunStatus.IN_PROGRESS,
        sa_column=Column(String(32), index=True),
    )
    current_floor: int = Field(default=1)
    current_question_index: int = Field(default=1)
    current_hp: int = Field(default=100)
    current_enemy_hp: int = Field(default=30)
    best_floor: int = Field(default=1)
    best_floor_reached_at: datetime
    started_at: datetime
    ended_at: Optional[datetime] = None


class DailyChallengeAnswer(SQLModel, table=True):
    __tablename__ = "daily_challenge_answer"
    __table_args__ = (
        UniqueConstraint("run_id", "question_id", name="uq_daily_answer_run_question"),
    )

    id: Optional[int] = Field(default=None, primary_key=True)
    run_id: int = Field(foreign_key="daily_challenge_run.id", index=True)
    floor: int = Field(index=True)
    question_index: int = Field(default=1)
    question_id: int = Field(foreign_key="question.id", index=True)
    answer_detail: dict[str, Any] = Field(default_factory=dict, sa_column=Column(_json_type()))
    is_correct: bool
    answered_at: datetime


class DailyChallengeUsedWord(SQLModel, table=True):
    __tablename__ = "daily_challenge_used_word"
    __table_args__ = (
        UniqueConstraint("day_id", "word_id", name="uq_daily_used_word"),
    )

    id: Optional[int] = Field(default=None, primary_key=True)
    day_id: int = Field(foreign_key="daily_challenge_day.id", index=True)
    floor: int = Field(index=True)
    question_index: int = Field(default=1)
    word_id: int = Field(foreign_key="word.id", index=True)
    created_at: datetime


class DailyChallengeQuestionRead(SQLModel):
    question_id: int
    floor: int
    question_index: int
    type: str
    content: dict[str, Any]


class DailyChallengeAnswerResultRead(SQLModel):
    is_correct: bool
    detail: dict[str, Any] = Field(default_factory=dict)


class DailyChallengeStateRead(SQLModel):
    run_id: int
    day_key: str
    status: str
    current_floor: int
    current_question_index: int
    current_hp: int
    current_enemy_hp: int
    player_max_hp: int
    player_attack: int
    enemy_max_hp: int
    enemy_attack: int
    heal_every_floors: int
    heal_amount: int
    best_floor: int
    best_floor_reached_at: datetime
    started_at: datetime
    ended_at: Optional[datetime] = None
    question: Optional[DailyChallengeQuestionRead] = None
    answer_result: Optional[DailyChallengeAnswerResultRead] = None


class DailyChallengeLeaderboardEntryRead(SQLModel):
    rank: int
    user_id: int
    username: str
    nickname: Optional[str] = None
    avatar_url: Optional[str] = None
    best_floor: int
    best_floor_reached_at: Optional[datetime] = None
    started_at: datetime
    status: str
    is_current_user: bool = False


class DailyChallengeOverviewRead(SQLModel):
    day_key: str
    opens_at: datetime
    closes_at: datetime
    seconds_until_reset: int
    user_status: str
    today_best_floor: Optional[int] = None
    active_run: Optional[DailyChallengeStateRead] = None
    leaderboard: list[DailyChallengeLeaderboardEntryRead] = Field(default_factory=list)
    current_user_rank: Optional[int] = None
    current_user_entry: Optional[DailyChallengeLeaderboardEntryRead] = None


class DailyChallengeAnswerSubmit(SQLModel):
    question_id: int
    selected_option: Optional[str] = None
    selected_sequence: Optional[list[str]] = None
    user_input: Optional[str] = None
    is_correct: Optional[bool] = None
    answer_detail: Optional[dict[str, Any]] = None
